compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2769,9 +2768,10 @@ impl<'tcx> ty::Instance<'tcx> {
                 // `src/test/ui/polymorphization/normalized_sig_types.rs`), and codegen not keeping
                 // track of a polymorphization `ParamEnv` to allow normalizing later.
                 let mut sig = match *ty.kind() {
-                    ty::FnDef(def_id, substs) => tcx
-                        .normalize_erasing_regions(tcx.param_env(def_id), tcx.bound_fn_sig(def_id))
-                        .subst(tcx, substs),
+                    ty::FnDef(def_id, substs) => tcx.normalize_erasing_regions(
+                        tcx.param_env(def_id),
+                        tcx.bound_fn_sig(def_id).subst(tcx, substs),
+                    ),
```

Comment:
and woops, wrong param env xd

Reply:
ok, it feels wrong but polymorphization breaks with this change, ah whatever.

Reply:
Any idea of a "proper" solution for polymorphism?

Reply:
well, two ideas:

1. don't normalize while computing whether we may polymorphize
2. only use the normalized types after polymorphization

i thinking about this, it would be correct to normalize twice here: once to get to the state assumed by polymorphization, and then once after the `subst` call to actually have a fully normalized type.

Actually, the result of `fn_sig_for_fn_abi` is only used by `fn_abi_new_uncached` which does the following :grin: 
```rust
        let sig = self.tcx.normalize_erasing_late_bound_regions(self.param_env, sig);
```

so we don't have to normalize here at all, except to deal with `2`. added a comment mentioning this.